pandas/core/common.py

```diff
@@ -1403,6 +1403,10 @@ def _fill_zeros(result, x, y, name, fill):
                 if np.isinf(fill):
                     np.putmask(result,(signs<0) & mask, -fill)
 
+                if "floordiv" in name:  # (PR 9308)
+                    nan_mask = ((y == 0) & (x == 0)).ravel()
+                    np.putmask(result, nan_mask, np.nan)
+
```

Comment:
@jreback, you asked if it's possible to not special case this.  I rewrote it to be more clean and in fact, as it's written now, you can get rid of the `if "floordiv" in name:` line, and it'll be fine.  It turns out that it'll only affect the floordiv operators' results anyway.  However, I think we should leave it as is for two reasons: 1) efficiency and (2) I think it makes it more readable since we really are special casing floordiv to ameliorate NumPy's strange treatment of floor division.

As for efficiency, we don't need to run these two lines for any other operator, for example,

    xx = pd.DataFrame(np.random.random((10000, 10000)))
    yy = xx / 0

The second line took 1.59s on average, while if you get rid of the if statement above, it takes 2.05s on average.